### PR TITLE
Refactor unit tests for WaitSteps

### PR DESCRIPTION
**Pull request by ivan-lednev**

### CODE REVIEW

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java [hunk 1]
```diff
@@ -441,4 +441,14 @@ void testWaitAlertDoesNotAppear()
...
     }
+    
+    @Test
+    void testWaitForElementDisappearance()
+    {
+    	By by = By.xpath(XPATH);
+    	IExpectedSearchContextCondition<?> condition = State.NOT_VISIBLE.getExpectedCondition(expectedSearchContextConditions, by);
+    	when(waitActions.wait(webElement, condition)).thenReturn(new WaitResult<>());
+    	waitSteps.waitForElementDisappearance(webElement, by);
+    	verify(waitActions).wait(webElement, condition);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: https://static.javadoc.io/org.mockito/mockito-core/3.0.0/org/mockito/Mockito.html#stubbing:
```java
 //Although it is possible to verify a stubbed invocation, usually it's just redundant
```

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java [hunk 2]
```diff
@@ -441,4 +441,14 @@ void testWaitAlertDoesNotAppear()
...
         verify(softAssert).assertFalse("Alert does not appear", waitResult.isWaitPassed());
     }
+    
+    @Test
+    void testWaitForElementDisappearance()
+    {
+    	By by = By.xpath(XPATH);
+    	IExpectedSearchContextCondition<?> condition = State.NOT_VISIBLE.getExpectedCondition(expectedSearchContextConditions, by);
+    	when(waitActions.wait(webElement, condition)).thenReturn(new WaitResult<>());
+    	waitSteps.waitForElementDisappearance(webElement, by);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: assert the result returned by the method

**Comment by EDbarvinsky** [comment]
change commit message to Add a unit test for a waiting step

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java [hunk 3]
```diff
@@ -112,25 +112,24 @@
     @Test
     void testElementByIdDisappears()
     {
+        By locator = By.xpath(".//*[normalize-space(@id)=\"id\"]");
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: By.xpath(".//*[normalize-space(@id)=\"id\"]") -> By.xpath(".//*[normalize-space(@id)='id']")

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java [hunk 4]
```diff
@@ -155,7 +154,9 @@ void testWaitTillElementAppears()
     {
         when(webUiContext.getSearchContext()).thenReturn(webElement);
         SearchAttributes attributes = new SearchAttributes(ActionAttributeType.ELEMENT_NAME, NAME);
-        when(waitActions.wait(webElement, expectedSearchActionsConditions.visibilityOfElement(attributes)))
+        IExpectedSearchContextCondition<WebElement> condition = mock(IExpectedSearchContextCondition.class);
+        when(expectedSearchActionsConditions.visibilityOfElement(attributes)).thenReturn(condition);
+        when(waitActions.wait(webElement, condition))
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: merge in one line

Comment (ikalinin1) [reply vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: here and everywhere; Line limit is 120 symbols

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java [hunk 5]
```diff
@@ -294,8 +306,10 @@ void testWaitTillElementWithTextDisappearsElementIsNotPresent()
     void testWaitTillElementWithXpathAppears()
     {
         when(webUiContext.getSearchContext()).thenReturn(webElement);
-        By by = By.xpath(".//xpath");
-        when(waitActions.wait(webElement, expectedSearchContextConditions.visibilityOfAllElementsLocatedBy(by)))
+        By by = By.xpath(XPATH);
```
Comment (ikalinin1) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: Make XPATH By type?

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java [hunk 6]
```diff
@@ -340,26 +355,32 @@ void testWaitTillElementWithXpathDisappears2()
...
-                .visibilityOfAllElementsLocatedBy(attributes))).thenReturn(new WaitResult<>());
-        waitSteps.waitDurationWithPollingDurationTillElementAppears(TIMEOUT, TIMEOUT, attributes);
+        IExpectedSearchContextCondition<List<WebElement>> condition = mock(IExpectedSearchContextCondition.class);
+        when(expectedSearchActionsConditions.visibilityOfAllElementsLocatedBy(attributes)).thenReturn(condition);
+        when(waitActions.wait(webElement, TIMEOUT, TIMEOUT, condition)).thenReturn(waitResult);
+        assertTrue(waitSteps.waitDurationWithPollingDurationTillElementAppears(TIMEOUT, TIMEOUT, attributes));
     }
 
     @Test
+    // TODO refactor. What needs to be refactored?
```
Comment (valfirst) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: please remove TODO-comments

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java (hunk 3 above)
Comment (ivan-lednev) [reply vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: The test fails if I change \"id\" to 'id'

**Comment by valfirst** [comment]
https://chris.beams.io/posts/git-commit/#imperative -> 5. Use the imperative mood in the subject line

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java [hunk 7]
```diff
@@ -441,4 +477,61 @@ void testWaitAlertDoesNotAppear()
...
+        WaitResult<Boolean> waitResult = new WaitResult<>();
+        when(waitActions.wait(webElement, TIMEOUT, condition)).thenReturn(waitResult);
+        waitResult.setWaitPassed(true);
+        assertTrue(waitSteps.waitForElementDisappearance(webElement, attributes, TIMEOUT));
+    }
+
+    @Test
+    void testWaitForElementDisappearanceByWithDuration()
+    {
+        By by = By.xpath(XPATH);
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: why not to extract it to separate static field?

Comment (valfirst) [reply vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/WaitStepsTests.java]: it's used several times, but not all tests require this field

**Pull request merged**